File: hermes-kotlin/src/main/kotlin/hermes/ingestion/Chunker.kt
package hermes.ingestion

import hermes.NodeType
import java.nio.file.Path

data class Chunk(
    val name: String,
    val nodeType: NodeType,
    val content: String,
    val startLine: Int,
    val endLine: Int,
    val summary: String
)

fun chunkFile(path: Path, content: String): List<Chunk> {
    val ext = path.toFile().extension
    return when (ext) {
        "rs" -> chunkRust(content)
        "kt", "kts", "java" -> chunkJvmLike(content)
        "md" -> chunkMarkdown(content)
        "tsx", "ts", "jsx", "js" -> chunkTypescript(content)
        else -> chunkWholeFile(path, content)
    }
}

// ── Rust chunker ──────────────────────────────────────────────────────────

private fun chunkRust(content: String): List<Chunk> {
    val chunks = mutableListOf<Chunk>()
    val lines = content.lines()
    var i = 0
    while (i < lines.size) {
        val line = lines[i].trim()
        val chunk = tryParseRustItem(line, lines, i)
        if (chunk != null) chunks.add(chunk)
        i++
    }
    return chunks
}

private fun tryParseRustItem(line: String, lines: List<String>, start: Int): Chunk? {
    val (name, nodeType) = when {
        line.startsWith("pub fn ") || line.startsWith("fn ") ||
        line.startsWith("pub async fn ") || line.startsWith("async fn ") ->
            (extractFnName(line) ?: return null) to NodeType.FUNCTION

        line.startsWith("pub struct ") || line.startsWith("struct ") ->
            (extractAfterKeyword(line, "struct") ?: return null) to NodeType.STRUCT

        line.startsWith("pub enum ") || line.startsWith("enum ") ->
            (extractAfterKeyword(line, "enum") ?: return null) to NodeType.ENUM

        line.startsWith("impl ") ->
            (extractImplName(line) ?: return null) to NodeType.IMPL

        line.startsWith("pub trait ") || line.startsWith("trait ") ->
            (extractAfterKeyword(line, "trait") ?: return null) to NodeType.TRAIT

        else -> return null
    }

    val end = findBlockEnd(lines, start)
    val blockContent = lines.subList(start, end + 1).joinToString("\n")
    val summary = buildSummary(name, nodeType, lines[start])

    return Chunk(
        name = name,
        nodeType = nodeType,
        content = blockContent,
        startLine = start + 1,
        endLine = end + 1,
        summary = summary
    )
}

// ── JVM-like (Kotlin/Java) chunker ───────────────────────────────────────

private fun chunkJvmLike(content: String): List<Chunk> {
    val chunks = mutableListOf<Chunk>()
    val lines = content.lines()

    for (i in lines.indices) {
        val trimmed = lines[i].trim()
        val (name, nodeType) = when {
            trimmed.contains("fun ") && trimmed.contains("(") ->
                (extractJvmFnName(trimmed) ?: continue) to NodeType.FUNCTION

            (trimmed.startsWith("class ") || trimmed.startsWith("data class ") ||
             trimmed.startsWith("open class ") || trimmed.startsWith("abstract class ") ||
             trimmed.contains(" class ")) && trimmed.contains("{") ->
                (extractAfterKeyword(trimmed, "class") ?: continue) to NodeType.STRUCT

            trimmed.startsWith("interface ") || trimmed.contains(" interface ") ->
                (extractAfterKeyword(trimmed, "interface") ?: continue) to NodeType.TRAIT

            trimmed.startsWith("enum class ") || trimmed.contains(" enum class ") ->
                (extractAfterKeyword(trimmed, "class") ?: continue) to NodeType.ENUM

            trimmed.startsWith("object ") || trimmed.contains(" object ") ->
                (extractAfterKeyword(trimmed, "object") ?: continue) to NodeType.MODULE

            else -> continue
        }

        val end = findBlockEnd(lines, i)
        val blockContent = lines.subList(i, end + 1).joinToString("\n")
        val summary = buildSummary(name, nodeType, lines[i])

        chunks.add(Chunk(
            name = name,
            nodeType = nodeType,
            content = blockContent,
            startLine = i + 1,
            endLine = end + 1,
            summary = summary
        ))
    }
    return chunks
}

private fun extractJvmFnName(line: String): String? {
    val afterFun = line.split("fun ").getOrNull(1) ?: return null
    val name = afterFun.split("(").firstOrNull()?.split("<")?.firstOrNull()?.trim()
    return if (name.isNullOrEmpty()) null else name
}

// ── Markdown chunker ─────────────────────────────────────────────────────

private fun chunkMarkdown(content: String): List<Chunk> {
    val chunks = mutableListOf<Chunk>()
    val lines = content.lines()
    var sectionStart: Pair<Int, String>? = null

    for (i in lines.indices) {
        val line = lines[i]
        if (line.startsWith("## ") || line.startsWith("# ")) {
            sectionStart?.let { (start, heading) ->
                chunks.add(Chunk(
                    name = heading,
                    nodeType = NodeType.DOCUMENT,
                    content = lines.subList(start, i).joinToString("\n"),
                    startLine = start + 1,
                    endLine = i,
                    summary = heading
                ))
            }
            sectionStart = i to line.trimStart('#').trim()
        }
    }
    sectionStart?.let { (start, heading) ->
        chunks.add(Chunk(
            name = heading,
            nodeType = NodeType.DOCUMENT,
            content = lines.subList(start, lines.size).joinToString("\n"),
            startLine = start + 1,
            endLine = lines.size,
            summary = heading
        ))
    }
    return chunks
}

// ── TypeScript/JS chunker ────────────────────────────────────────────────

private fun chunkTypescript(content: String): List<Chunk> {
    val chunks = mutableListOf<Chunk>()
    val lines = content.lines()

    for (i in lines.indices) {
        val trimmed = lines[i].trim()
        if (isTsFunctionStart(trimmed) || isTsComponentStart(trimmed)) {
            val name = extractTsName(trimmed) ?: "anonymous_$i"
            val end = findBlockEnd(lines, i)
            chunks.add(Chunk(
                name = name,
                nodeType = NodeType.FUNCTION,
                content = lines.subList(i, end + 1).joinToString("\n"),
                startLine = i + 1,
                endLine = end + 1,
                summary = "TypeScript function: $name"
            ))
        }
    }
    return chunks
}

// ── Whole-file fallback ──────────────────────────────────────────────────

private fun chunkWholeFile(path: Path, content: String): List<Chunk> {
    val name = path.fileName.toString()
    return listOf(Chunk(
        name = name,
        nodeType = NodeType.FILE,
        content = content,
        startLine = 1,
        endLine = content.lines().size,
        summary = "File: $name"
    ))
}

// ── Helpers ──────────────────────────────────────────────────────────────

private fun extractFnName(line: String): String? {
    val afterFn = line.split("fn ").getOrNull(1) ?: return null
    val name = afterFn.split("(").firstOrNull()?.split("<")?.firstOrNull()?.trim()
    return if (name.isNullOrEmpty()) null else name
}

private fun extractAfterKeyword(line: String, keyword: String): String? {
    val after = line.split("$keyword ").getOrNull(1) ?: return null
    val name = after.split("{").firstOrNull()
        ?.split("<")?.firstOrNull()
        ?.split("(")?.firstOrNull()
        ?.split(":")?.firstOrNull()
        ?.trim()
    return if (name.isNullOrEmpty()) null else name
}

private fun extractImplName(line: String): String? {
    val afterImpl = line.removePrefix("impl ").ifEmpty { return null }
    val name = afterImpl.split("{").firstOrNull()
        ?.split("for ")?.lastOrNull()
        ?.split("<")?.firstOrNull()
        ?.trim()
    return if (name.isNullOrEmpty()) null else name
}

private fun findBlockEnd(lines: List<String>, start: Int): Int {
    var depth = 0
    var foundOpen = false

    for (i in start until lines.size) {
        for (ch in lines[i]) {
            if (ch == '{') { depth++; foundOpen = true }
            else if (ch == '}') depth--
        }
        if (foundOpen && depth <= 0) return i
    }
    return (start + 1).coerceAtMost(lines.size - 1)
}

private fun buildSummary(name: String, nodeType: NodeType, firstLine: String): String {
    val cleanLine = firstLine.trim()
    return if (cleanLine.length > 80) "${nodeType.value}: $name"
    else "${nodeType.value}: $cleanLine"
}

private fun isTsFunctionStart(line: String): Boolean {
    return (line.startsWith("export function ") ||
            line.startsWith("function ") ||
            line.startsWith("export const ") ||
            line.startsWith("const ")) &&
            (line.contains("=>") || line.contains("("))
}

private fun isTsComponentStart(line: String): Boolean {
    return line.startsWith("export default function ") || line.startsWith("export default class ")
}

private fun extractTsName(line: String): String? {
    for (keyword in listOf("function ", "const ", "class ")) {
        val after = line.split(keyword).getOrNull(1) ?: continue
        val name = after.split("(").firstOrNull()
            ?.split("=")?.firstOrNull()
            ?.split(":")?.firstOrNull()
            ?.split("<")?.firstOrNull()
            ?.trim()
        if (!name.isNullOrEmpty()) return name
    }
    return null
}

File: src/ingestion/chunker.rs
use crate::graph::NodeType;
use std::path::Path;

#[derive(Debug, Clone)]
pub struct Chunk {
    pub name: String,
    pub node_type: NodeType,
    pub content: String,
    pub start_line: usize,
    pub end_line: usize,
    pub summary: String,
}

pub fn chunk_file(path: &Path, content: &str) -> Vec<Chunk> {
    let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("");

    match ext {
        "rs" => chunk_rust(content),
        "md" => chunk_markdown(content),
        "tsx" | "ts" | "jsx" | "js" => chunk_typescript(content),
        _ => chunk_whole_file(path, content),
    }
}

fn chunk_rust(content: &str) -> Vec<Chunk> {
    let mut chunks = Vec::new();
    let lines: Vec<&str> = content.lines().collect();
    let mut i = 0;

    while i < lines.len() {
        let line = lines[i].trim();

        if let Some(chunk) = try_parse_rust_item(line, &lines, i) {
            chunks.push(chunk);
        }
        i += 1;
    }

    chunks
}

fn try_parse_rust_item(line: &str, lines: &[&str], start: usize) -> Option<Chunk> {
    let (name, node_type) = if line.starts_with("pub fn ")
        || line.starts_with("fn ")
        || line.starts_with("pub async fn ")
        || line.starts_with("async fn ")
    {
        (extract_fn_name(line)?, NodeType::Function)
    } else if line.starts_with("pub struct ") || line.starts_with("struct ") {
        (extract_after_keyword(line, "struct")?, NodeType::Struct)
    } else if line.starts_with("pub enum ") || line.starts_with("enum ") {
        (extract_after_keyword(line, "enum")?, NodeType::Enum)
    } else if line.starts_with("impl ") {
        (extract_impl_name(line)?, NodeType::Impl)
    } else if line.starts_with("pub trait ") || line.starts_with("trait ") {
        (extract_after_keyword(line, "trait")?, NodeType::Trait)
    } else {
        return None;
    };

    let end = find_block_end(lines, start);
    let block_content: String = lines[start..=end].join("\n");
    let summary = build_summary(&name, &node_type, lines[start]);

    Some(Chunk {
        name,
        node_type,
        content: block_content,
        start_line: start + 1,
        end_line: end + 1,
        summary,
    })
}

fn chunk_markdown(content: &str) -> Vec<Chunk> {
    let mut chunks = Vec::new();
    let lines: Vec<&str> = content.lines().collect();
    let mut section_start: Option<(usize, String)> = None;

    for (i, line) in lines.iter().enumerate() {
        if line.starts_with("## ") || line.starts_with("# ") {
            if let Some((start, heading)) = section_start.take() {
                let section_content = lines[start..i].join("\n");
                chunks.push(Chunk {
                    name: heading.clone(),
                    node_type: NodeType::Document,
                    content: section_content,
                    start_line: start + 1,
                    end_line: i,
                    summary: heading,
                });
            }
            section_start = Some((i, line.trim_start_matches('#').trim().to_string()));
        }
    }

    if let Some((start, heading)) = section_start {
        let section_content = lines[start..].join("\n");
        chunks.push(Chunk {
            name: heading.clone(),
            node_type: NodeType::Document,
            content: section_content,
            start_line: start + 1,
            end_line: lines.len(),
            summary: heading,
        });
    }

    chunks
}

fn chunk_typescript(content: &str) -> Vec<Chunk> {
    let mut chunks = Vec::new();
    let lines: Vec<&str> = content.lines().collect();

    for (i, line) in lines.iter().enumerate() {
        let trimmed = line.trim();
        if is_ts_function_start(trimmed) || is_ts_component_start(trimmed) {
            let name = extract_ts_name(trimmed).unwrap_or_else(|| format!("anonymous_{i}"));
            let end = find_block_end(&lines, i);
            let block_content = lines[i..=end].join("\n");
            chunks.push(Chunk {
                name: name.clone(),
                node_type: NodeType::Function,
                content: block_content,
                start_line: i + 1,
                end_line: end + 1,
                summary: format!("TypeScript function: {name}"),
            });
        }
    }

    chunks
}

fn chunk_whole_file(path: &Path, content: &str) -> Vec<Chunk> {
    let name = path
        .file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_else(|| "unknown".to_string());

    vec![Chunk {
        name: name.clone(),
        node_type: NodeType::File,
        content: content.to_string(),
        start_line: 1,
        end_line: content.lines().count(),
        summary: format!("File: {name}"),
    }]
}

fn extract_fn_name(line: &str) -> Option<String> {
    let after_fn = line.split("fn ").nth(1)?;
    let name = after_fn.split('(').next()?.split('<').next()?.trim();
    if name.is_empty() {
        None
    } else {
        Some(name.to_string())
    }
}

fn extract_after_keyword(line: &str, keyword: &str) -> Option<String> {
    let kw_with_space = format!("{keyword} ");
    let after = line.split(&kw_with_space).nth(1)?;
    let name = after
        .split('{')
        .next()?
        .split('<')
        .next()?
        .split('(')
        .next()?
        .trim();
    if name.is_empty() {
        None
    } else {
        Some(name.to_string())
    }
}

fn extract_impl_name(line: &str) -> Option<String> {
    let after_impl = line.strip_prefix("impl ")?;
    let name = after_impl
        .split('{')
        .next()?
        .split("for ")
        .last()?
        .split('<')
        .next()?
        .trim();
    if name.is_empty() {
        None
    } else {
        Some(name.to_string())
    }
}

fn find_block_end(lines: &[&str], start: usize) -> usize {
    let mut depth: i32 = 0;
    let mut found_open = false;

    for (i, line) in lines.iter().enumerate().skip(start) {
        for ch in line.chars() {
            if ch == '{' {
                depth += 1;
                found_open = true;
            } else if ch == '}' {
                depth -= 1;
            }
        }
        if found_open && depth <= 0 {
            return i;
        }
    }

    (start + 1).min(lines.len() - 1)
}

fn build_summary(name: &str, node_type: &NodeType, first_line: &str) -> String {
    let type_str = node_type.as_str();
    let clean_line = first_line.trim();
    if clean_line.len() > 80 {
        format!("{type_str}: {name}")
    } else {
        format!("{type_str}: {clean_line}")
    }
}

fn is_ts_function_start(line: &str) -> bool {
    (line.starts_with("export function ")
        || line.starts_with("function ")
        || line.starts_with("export const ")
        || line.starts_with("const "))
        && (line.contains("=>") || line.contains("("))
}

fn is_ts_component_start(line: &str) -> bool {
    line.starts_with("export default function ") || line.starts_with("export default class ")
}

fn extract_ts_name(line: &str) -> Option<String> {
    for keyword in &["function ", "const ", "class "] {
        if let Some(after) = line.split(keyword).nth(1) {
            let name = after
                .split('(')
                .next()?
                .split('=')
                .next()?
                .split(':')
                .next()?
                .split('<')
                .next()?
                .trim();
            if !name.is_empty() {
                return Some(name.to_string());
            }
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn chunk_rust_function() {
        let code = "pub fn hello(name: &str) -> String {\n    format!(\"Hello {name}\")\n}\n";
        let chunks = chunk_rust(code);
        assert_eq!(chunks.len(), 1);
        assert_eq!(chunks[0].name, "hello");
        assert_eq!(chunks[0].node_type, NodeType::Function);
    }

    #[test]
    fn chunk_rust_struct() {
        let code = "pub struct Config {\n    pub port: u16,\n}\n";
        let chunks = chunk_rust(code);
        assert_eq!(chunks.len(), 1);
        assert_eq!(chunks[0].name, "Config");
        assert_eq!(chunks[0].node_type, NodeType::Struct);
    }

    #[test]
    fn chunk_markdown_sections() {
        let md = "# Title\nIntro\n## Section A\nContent A\n## Section B\nContent B\n";
        let chunks = chunk_markdown(md);
        assert_eq!(chunks.len(), 3);
        assert_eq!(chunks[0].name, "Title");
    }

    #[test]
    fn extract_fn_name_variants() {
        assert_eq!(extract_fn_name("pub fn hello()"), Some("hello".to_string()));
        assert_eq!(extract_fn_name("fn main()"), Some("main".to_string()));
        assert_eq!(
            extract_fn_name("pub async fn fetch_data()"),
            Some("fetch_data".to_string())
        );
    }

    #[test]
    fn chunk_rust_enum() {
        let code = "pub enum Status {\n    Active,\n    Inactive,\n}\n";
        let chunks = chunk_rust(code);
        assert_eq!(chunks.len(), 1);
        assert_eq!(chunks[0].name, "Status");
        assert_eq!(chunks[0].node_type, NodeType::Enum);
    }

    #[test]
    fn chunk_rust_impl_block() {
        let code = "impl MyStruct {\n    pub fn method(&self) {}\n}\n";
        let chunks = chunk_rust(code);
        // impl block itself is a chunk
        let impl_chunk = chunks.iter().find(|c| c.node_type == NodeType::Impl);
        assert!(impl_chunk.is_some());
        assert_eq!(impl_chunk.unwrap().name, "MyStruct");
    }

    #[test]
    fn chunk_rust_trait() {
        let code = "pub trait Searchable {\n    fn search(&self) -> Vec<String>;\n}\n";
        let chunks = chunk_rust(code);
        // The chunker picks up both the trait block and the fn declaration inside it
        assert!(!chunks.is_empty());
        let trait_chunk = chunks.iter().find(|c| c.node_type == NodeType::Trait);
        assert!(trait_chunk.is_some(), "expected a Trait chunk");
        assert_eq!(trait_chunk.unwrap().name, "Searchable");
    }

    #[test]
    fn extract_impl_name_simple() {
        assert_eq!(extract_impl_name("impl MyStruct {"), Some("MyStruct".to_string()));
    }

    #[test]
    fn extract_impl_name_for_trait() {
        assert_eq!(
            extract_impl_name("impl Display for MyStruct {"),
            Some("MyStruct".to_string())
        );
    }

    #[test]
    fn chunk_file_dispatch_rust() {
        use std::path::PathBuf;
        let path = PathBuf::from("src/lib.rs");
        let code = "pub fn run() {\n    println!(\"hi\");\n}\n";
        let chunks = chunk_file(&path, code);
        assert!(!chunks.is_empty());
        assert_eq!(chunks[0].node_type, NodeType::Function);
    }

    #[test]
    fn chunk_file_dispatch_markdown() {
        use std::path::PathBuf;
        let path = PathBuf::from("README.md");
        let md = "# Overview\nIntro\n## Usage\nDetails\n";
        let chunks = chunk_file(&path, md);
        assert!(!chunks.is_empty());
        assert_eq!(chunks[0].node_type, NodeType::Document);
    }

    #[test]
    fn chunk_file_dispatch_unknown_extension() {
        use std::path::PathBuf;
        let path = PathBuf::from("config.toml");
        let content = "[package]\nname = \"hermes\"\n";
        let chunks = chunk_file(&path, content);
        assert_eq!(chunks.len(), 1);
        assert_eq!(chunks[0].node_type, NodeType::File);
        assert_eq!(chunks[0].name, "config.toml");
    }

    #[test]
    fn chunk_typescript_function() {
        let code = "export function handleRequest(req: Request) {\n    return req;\n}\n";
        let chunks = chunk_typescript(code);
        assert!(!chunks.is_empty());
        assert_eq!(chunks[0].name, "handleRequest");
        assert_eq!(chunks[0].node_type, NodeType::Function);
    }

    #[test]
    fn chunk_typescript_arrow_const() {
        let code = "const fetchData = async (url: string) => {\n    return fetch(url);\n};\n";
        let chunks = chunk_typescript(code);
        assert!(!chunks.is_empty());
        assert_eq!(chunks[0].name, "fetchData");
    }

    #[test]
    fn build_summary_short_line() {
        let summary = build_summary("my_fn", &NodeType::Function, "pub fn my_fn() {");
        assert_eq!(summary, "function: pub fn my_fn() {");
    }

    #[test]
    fn build_summary_long_line() {
        let long_line = "pub fn a_very_long_function_name_that_exceeds_eighty_characters_limit_for_sure(x: u32) {";
        let summary = build_summary("a_very_long_function_name_that_exceeds_eighty_characters_limit_for_sure", &NodeType::Function, long_line);
        assert_eq!(summary, "function: a_very_long_function_name_that_exceeds_eighty_characters_limit_for_sure");
    }

    #[test]
    fn chunk_whole_file_produces_single_chunk() {
        use std::path::PathBuf;
        let path = PathBuf::from("data.json");
        let content = "{\"key\": \"value\"}";
        let chunks = chunk_whole_file(&path, content);
        assert_eq!(chunks.len(), 1);
        assert_eq!(chunks[0].name, "data.json");
        assert_eq!(chunks[0].start_line, 1);
    }

    #[test]
    fn markdown_single_section() {
        let md = "# Only One\nSome content here\n";
        let chunks = chunk_markdown(md);
        assert_eq!(chunks.len(), 1);
        assert_eq!(chunks[0].name, "Only One");
    }

    #[test]
    fn markdown_empty_returns_empty() {
        let chunks = chunk_markdown("");
        assert!(chunks.is_empty());
    }
}
